### PR TITLE
Remove hard coded codes from workflow

### PR DESCRIPTION
**Pull request by superstar54**
The codes (`pw_code`, `dos_code`, `projwfc_code`) are hard coded in the builder, which is not good for the plugin design. This PR uses `get_selected_codes`. 
The `get_resources` method is also used to make the code more modularized.

### CODE REVIEW

File: src/aiidalab_qe/app/submission/__init__.py [hunk 1]
```diff
@@ -412,30 +433,23 @@ def _create_builder(self) -> ProcessBuilderNamespace:
...
-        pw_code = self.pw_code.value
-        dos_code = self.dos_code.value
-        projwfc_code = self.projwfc_code.value
-
-        parameters = deepcopy(self.input_parameters)
+        self.builder_parameters = deepcopy(self.input_parameters)
+        # add codes info into input_parameters
+        self.builder_parameters["codes"] = self.get_selected_codes()
+        self.builder_parameters["resources"] = self.get_resources()
+        #
```
Comment (unkcpz) [inline src/aiidalab_qe/app/submission/__init__.py:440]:
```suggestion

```

File: src/aiidalab_qe/app/submission/__init__.py [hunk 2]
```diff
@@ -346,6 +359,14 @@ def _on_submit_button_clicked(self, _):
         self.submit_button.disabled = True
         self.submit()
 
+    def get_selected_codes(self):
+        codes = {
```
Comment (unkcpz) [inline src/aiidalab_qe/app/submission/__init__.py:363]:
```suggestion
        """Return the dict of code name as key and code node UUID as value"""
        codes = {
```

File: src/aiidalab_qe/app/submission/__init__.py [hunk 3]
```diff
@@ -346,6 +359,14 @@ def _on_submit_button_clicked(self, _):
...
+            "pw_code": self.pw_code.value,
+            "dos_code": self.dos_code.value,
+            "projwfc_code": self.projwfc_code.value,
```
Comment (unkcpz) [inline src/aiidalab_qe/app/submission/__init__.py:366]:
```suggestion
            "pw": self.pw_code.value,
            "dos": self.dos_code.value,
            "projwfc": self.projwfc_code.value,
```

File: src/aiidalab_qe/workflows/__init__.py [hunk 4]
```diff
@@ -134,7 +132,7 @@ def get_builder_from_protocol(
         relax_overrides = {"base": parameters["advanced"]}
         protocol = parameters["workchain"]["protocol"]
         relax_builder = PwRelaxWorkChain.get_builder_from_protocol(
-            code=pw_code,
+            code=codes.get("pw_code"),
```
Comment (unkcpz) [inline src/aiidalab_qe/workflows/__init__.py:135]:
```suggestion
            code=codes.get("pw"),
```

**Review by unkcpz — CHANGES_REQUESTED** [review]
Thanks! Just minor requests. @superstar54

**Comment by superstar54** [comment]
@unkcpz I changed as you suggested. Since we replaced the `pw_code` with `pw`, the `qeapp.yaml` and the workchains are also updated.

File: src/aiidalab_qe/app/submission/__init__.py [hunk 5]
```diff
@@ -226,14 +226,16 @@ def _toggle_install_widgets(self, change):
...
-                "pw_code",
-                "dos_code",
-                "projwfc_code",
+                "pw",
+                "dos",
+                "projwfc",
             ]:
                 try:
-                    code_widget = getattr(self, code)
+                    code_widget = getattr(self, f"{code}_code")
```
Comment (unkcpz) [inline src/aiidalab_qe/app/submission/__init__.py:234]: Ah, I didn't notice this is directly used in `getattr`. Then the original code is good enough.

**Review by unkcpz — APPROVED** [review]
Thanks! Looks all good. 

I just merge https://github.com/aiidalab/aiidalab-qe/pull/449, and what to use it as a test. So I'll rebase this PR and see if from PR everything is built correctly. Hope you don't mind.

**Comment by unkcpz** [comment]
Hi @superstar54, after rebasing the tests failed. But I compared the two branches and didn't find the difference. Do you have any clue why it is happening, can you check the changes are all correct?

**Pull request merged**